### PR TITLE
fix(react): add `<ClientOnly/>` internally

### PR DESCRIPTION
**Pull request by Collection50**
related #528 

# Overview

Implemented the `ClientOnly` component.

## PR Checklist

- [x] I did below actions if need

1. I read the [Contributing Guide](https://github.com/toss/suspensive/blob/main/CONTRIBUTING.md)
2. I added documents and tests.

### CODE REVIEW

File: packages/react/src/components/useIsClient.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+}
+
+function getServerSnapshot() {
+  return true
+}
+
+// hook interface instead of useIsClient in hooks
+export function useIsClient() {
+  return useSyncExternalStore(() => noop, getSnapshot, getServerSnapshot)
+}
```
Comment (manudeli) [inline packages/react/src/components/useIsClient.ts:15]: This change is for next stage. Use [as-is useIsClient](https://github.com/toss/suspensive/blob/main/packages/react/src/hooks/useIsClient.ts) please first for current user.

File: packages/react/src/components/ClientOnly.tsx [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import { type SuspenseProps } from 'react'
+import { useIsClient } from './useIsClient'
+
+export const ClientOnly = ({ children, fallback }: SuspenseProps) => {
+  const isClient = useIsClient()
+
+  return isClient ? <>{children}</> : <>{fallback}</>
+}
```
Comment (manudeli) [inline packages/react/src/components/ClientOnly.tsx:8]: 1. Shortly if we can please
2. This is ClientOnly. so make type ClientOnlyProps please

```suggestion
export const ClientOnly = ({ children, fallback }: ClientOnlyProps) => <>{useIsClient() ? children : fallback}</>
```

File: packages/react/src/Suspense.tsx [hunk 3]
```diff
@@ -1,10 +1,13 @@
...
-import { useIsClient } from './hooks'
 import type { PropsWithDevMode } from './utility-types'
 
-const SuspenseClientOnly = (props: ReactSuspenseProps) =>
-  useIsClient() ? <ReactSuspense {...props} /> : <>{props.fallback}</>
+const SuspenseClientOnly = (props: ReactSuspenseProps) => (
+  <ClientOnly fallback={props.fallback}>
+    <ReactSuspense {...props} />
+  </ClientOnly>
+)
```
Comment (manudeli) [inline packages/react/src/Suspense.tsx:10]: 👍

**Review by manudeli — CHANGES_REQUESTED** [review]

File: packages/react/src/components/ClientOnly.tsx (hunk 2 above)
Comment (Collection50) [reply packages/react/src/components/ClientOnly.tsx:8]: I think most of the time `<ClientOnly/>` be used as the HOC, so it need a `children`.
What do you think about using it as?

```ts
interface ClienOnlyProp {
  children: ReactNode:
  fallback?: ReactNode
}
```

Comment (manudeli) [reply packages/react/src/components/ClientOnly.tsx:8]: Cool! That's good for me

File: packages/react/src/components/useIsClient.ts (hunk 1 above)
Comment (Collection50) [reply packages/react/src/components/useIsClient.ts:15]: How about this?

```ts
import React, { useState } from 'react'
import { noop } from '../utils'
import { useIsomorphicLayoutEffect } from './useIsomorphicLayoutEffect'

export const useIsClient = () => {
  const [isClient, setIsClient] = useState(false)

  useIsomorphicLayoutEffect(() => {
    setIsClient(true)
  }, [])

  if (typeof React['useSyncExternalStore'] === 'function') {
    return React['useSyncExternalStore'](
      () => noop,
      () => true,
      () => false
    )
  }

  return isClient
}
```

Comment (Collection50) [reply packages/react/src/components/useIsClient.ts:15]: This is a variant of the existing `useIsClient` and does not change the interface of the current `Suspensive` user; rather, the migration to the next stage is **much more acceptable**.

Comment (manudeli) [reply packages/react/src/components/useIsClient.ts:15]: > This is a variant of the existing useIsClient and does not change the interface of the current Suspensive user; rather, the migration to the next stage is much more acceptable.

I agree your opinion. but we should change our feature more gradually. I want to check this feature change as beta version. use original useIsClient please. In my opinion, This pull request should be just only for adding ClientOnly component. not improving Suspense clientOnly.

Comment (Collection50) [reply packages/react/src/components/useIsClient.ts:15]: Thank you for your opinion. I'll reflect it.

Comment (manudeli) [reply packages/react/src/components/useIsClient.ts]: Thanks for accepting my opinion. I also accept your opinion if we can too. but we should be more careful for this library users. 👍

Comment (Collection50) [reply packages/react/src/components/useIsClient.ts]: I'm just grateful to be able to contribute to a great library, with a developer I admire. If there's anything else I can contribute to, I'd love to.

File: packages/react/src/components/ClientOnly.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    render(
+      <ClientOnly fallback={<div>Loading...</div>}>
+        <div>Client Content</div>
+      </ClientOnly>
+    )
+
+    expect(screen.getByText('Loading...')).toBeInTheDocument()
+    expect(screen.queryByText('Client Content')).not.toBeInTheDocument()
+  })
+})
```
Comment (manudeli) [inline packages/react/src/components/ClientOnly.spec.tsx:39]: Cool test code!

**Review by manudeli — APPROVED** [review]
Thanks! @Collection50

**Comment by manudeli** [comment]
@Collection50 resolve ci:type error please.

**Review by manudeli — APPROVED** [review]
@Collection50 Let's improve clientOnly gradually after this pull request merging

**Pull request merged**